go/interpreter/lfvm/instructions_test.go

```diff
@@ -880,3 +880,147 @@ func TestExpansionCostOverflow(t *testing.T) {
+			refund := selfDestructRefund(test.destructed, test.revision)
+			if refund != test.refund {
+				t.Errorf("unexpected refund, wanted %d, got %d", test.refund, refund)
+			}
+		})
+	}
+}
+
+func TestSelfDestruct_BeneficiaryAccessCost(t *testing.T) {
+	tests := map[string]struct {
+		accessStatus tosca.AccessStatus
+		gas          tosca.Gas
+	}{
+		"cold account": {
+			accessStatus: tosca.ColdAccess,
+			gas:          2_600,
+		},
+		"warm account": {
+			accessStatus: tosca.WarmAccess,
+			gas:          0,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			gas := selfDestructedBeneficiaryAccessCost(test.accessStatus)
+			if gas != test.gas {
+				t.Errorf("unexpected gas, wanted %d, got %d", test.gas, gas)
+			}
+		})
+	}
+}
+
+func TestSelfDestruct_NewAccountCost(t *testing.T) {
+
+	tests := map[string]struct {
+		beneficiaryExists bool
+		balance           tosca.Value
+		gas               tosca.Gas
+	}{
+		"account exists no balance": {
+			beneficiaryExists: true,
+			balance:           tosca.Value{},
+			gas:               0,
+		},
+		"account exists with balance": {
+			beneficiaryExists: true,
+			balance:           tosca.Value{1},
+			gas:               0,
+		},
+		"new account without balance": {
+			beneficiaryExists: false,
+			balance:           tosca.Value{},
+			gas:               0,
+		},
+		"new account with balance": {
+			beneficiaryExists: false,
+			balance:           tosca.Value{1},
+			gas:               25_000,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			cost := selfDestructNewAccountCost(test.beneficiaryExists, test.balance)
+			if cost != test.gas {
+				t.Errorf("unexpected gas, wanted %d, got %d", test.gas, cost)
+			}
+		})
+	}
+}
+
+func TestSelfDestruct_ExistingAccountToNewBeneficiary(t *testing.T) {
+	// This tests produces the combination of context calls/results for the maximum dynamic gas cost possible.
+
+	beneficiaryAddress := tosca.Address{1}
+	selfAddress := tosca.Address{2}
+
+	ctrl := gomock.NewController(t)
+	runContext := tosca.NewMockRunContext(ctrl)
+	runContext.EXPECT().AccessAccount(beneficiaryAddress).Return(tosca.ColdAccess)
+	runContext.EXPECT().AccountExists(beneficiaryAddress).Return(false)
+	runContext.EXPECT().GetBalance(selfAddress).Return(tosca.Value{1})
+	runContext.EXPECT().SelfDestruct(selfAddress, beneficiaryAddress).Return(true)
+
+	ctxt := context{
+		params: tosca.Parameters{
+			BlockParameters: tosca.BlockParameters{
+				Revision: tosca.R13_Cancun,
+			},
+			Recipient: selfAddress,
+		},
+		status:  statusRunning,
+		stack:   NewStack(),
+		memory:  NewMemory(),
+		context: runContext,
+		// 25_000 for new account, 2_600 for beneficiary access
+		gas: 27_600,
+	}
+	ctxt.stack.push(new(uint256.Int).SetBytes(beneficiaryAddress[:]))
+
+	opSelfdestruct(&ctxt)
+
+	if ctxt.status != statusSelfDestructed {
+		t.Errorf("unexpected status, wanted %v, got %v", statusRunning, ctxt.status)
+	}
+	if ctxt.gas != 0 {
```

Comment:
For defensive coding purposes I would add a small positive `delta` to the gas and check that this `delta` is present after, just to make sure that the operation is not generally setting the remaining gas to 0.